lib/rules/no-mutation-props.js

```diff
@@ -0,0 +1,227 @@
+/**
+ * @fileoverview Prevent mutation of this.props
+ * @author Ian Schmitz, Joey Baker
+ */
+'use strict';
+
+const Components = require('../util/Components');
+const ARRAY_MUTATIONS = ['push', 'pop', 'shift', 'unshift', 'copyWithin', 'fill', 'reverse', 'sort', 'splice'];
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Prevent mutation of this.props',
+      category: 'Possible Errors',
+      recommended: false
+    }
+  },
+
+  create: Components.detect((context, components, utils) => {
+    /**
+     * Gets a unique ID for a node
+     * @param {Object} node The node to get an ID for
+     */
+    const getId = node => node && node.range.join(':');
+
+    const propsNodes = {};
+    /**
+     * Adds a node to the list of nodes that come from this.props
+     * @param {Object} node The node to add
+     */
+    const addToPropsNodes = node => {
+      const id = getId(node);
+
+      if (!id) {
+        return null;
+      }
+
+      if (!propsNodes[id]) {
+        propsNodes[id] = node;
+      }
+      return propsNodes[id];
+    };
+
+    /**
+     * Gets a node if it's been declared as a propsNode
+     * @param {Object} node The node to get
+     */
+    const getPropsNode = node => propsNodes[getId(node)];
+
+    /**
+     * Determines if a given node is from `this.props`
+     * @param {Object} node The node to examine
+     */
+    function isPropsNode (node) {
+      if (!node) {
+        return false;
+      }
+
+      let topObject = node;
+
+      // when looking at an object path, get the top-most object
```

Comment:
this might be helpful to move to a helper method, so that this function's code can be simpler.

Reply:
I hear you, but since this is only used in one place, I'm not sure the abstraction is worth it.